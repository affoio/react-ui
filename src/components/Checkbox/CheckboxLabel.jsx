import styled from 'styled-components';
import { getTypographyFontSizeStyles } from '../../theme/selectors';

const CheckboxLabel = styled.div`
  display: inline-block;
  ${({ theme }) => getTypographyFontSizeStyles(theme, 300)};
`;

export default CheckboxLabel;
