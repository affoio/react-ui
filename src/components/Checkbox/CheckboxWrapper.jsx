import styled from 'styled-components';
import { getMultipliedIndent } from '../../theme/selectors';
import { addStyleIfPropTruly } from '../../helpers/styledComponents';

const CheckboxWrapper = styled.div`
  display: inline-flex;
  align-items: center;
  height: 24px;
  margin-right: ${({ theme }) => getMultipliedIndent(theme, 2)};
  &:hover {
    cursor: pointer;
  }

  ${addStyleIfPropTruly(
    'disabled',
    `
    opacity: 0.5;

    &:hover {
      cursor: default;
    }
  `
  )};
`;

export default CheckboxWrapper;
