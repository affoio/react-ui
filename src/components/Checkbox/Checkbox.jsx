import React, { useState } from 'react';
import styled, { css } from 'styled-components';
import { bool, string, func, oneOfType, element } from 'prop-types';
import colors from '../../colors';

import CheckboxWrapper from './CheckboxWrapper';
import CheckboxLabel from './CheckboxLabel';
import { getMultipliedIndentFromProp } from '../../theme/selectors';
import check from './img/check.svg';
import { addStyleIfPropTruly } from '../../helpers/styledComponents';

const CheckboxStyled = styled.div`
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid ${colors.indigoBlue[800]};
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: ${getMultipliedIndentFromProp(2)};
  transition: background-color 0.3s ease-out;

  ${({ checked, disabled }) =>
    !checked &&
    !disabled &&
    css`
      ${CheckboxWrapper}:hover & {
        background-color: ${colors.indigoBlue[200]};
      }
    `}

  ${addStyleIfPropTruly(
    'checked',
    `
      background-color: ${colors.indigoBlue[800]};
      background-image: url(${check});
      background-repeat: no-repeat;
      background-position: center;
    `
  )}
  ${addStyleIfPropTruly('error', `border: 2px solid ${colors.cerise[800]};`)}
`;

const propTypes = {
  defaultChecked: bool,
  label: oneOfType([string, element]),
  disabled: bool,
  onChange: func,
  error: oneOfType([string, bool]),
};

const Checkbox = ({ defaultChecked = false, label, disabled, onChange, error }) => {
  const [checked, changeCheck] = useState(defaultChecked);
  const checkboxClickHandler = () => {
    if (!disabled) {
      changeCheck(!checked);
      if (onChange) {
        onChange(!checked);
      }
    }
  };

  return (
    <CheckboxWrapper onClick={checkboxClickHandler} disabled={disabled}>
      <CheckboxStyled {...{ checked, disabled, error }} />
      {label && <CheckboxLabel>{label}</CheckboxLabel>}
    </CheckboxWrapper>
  );
};

Checkbox.propTypes = propTypes;

export default Checkbox;
