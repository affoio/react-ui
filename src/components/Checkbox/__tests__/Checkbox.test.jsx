import React from 'react';
import { render, mount } from 'enzyme';
import Checkbox from '../Checkbox';

describe('<Checkbox />', () => {
  it('renders correctly', () => {
    const tree = render(<Checkbox />);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with label', () => {
    const tree = render(<Checkbox label="Label" />);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with defaultChecked', () => {
    const tree = render(<Checkbox defaultChecked />);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with disabled', () => {
    const tree = render(<Checkbox disabled />);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly after check', () => {
    const wrapper = mount(<Checkbox />);
    wrapper.simulate('click');

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('correctly calls onChange', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<Checkbox onChange={mockFn} />);
    wrapper.simulate('click');
    expect(mockFn).toBeCalledWith(true);
    wrapper.simulate('click');
    expect(mockFn).toBeCalledWith(false);
  });

  it('doesnt calls onChange when disabled', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<Checkbox disabled onChange={mockFn} />);
    wrapper.simulate('click');
    expect(mockFn).not.toBeCalled();
  });
});
