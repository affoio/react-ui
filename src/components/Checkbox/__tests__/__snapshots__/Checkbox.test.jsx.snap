// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders correctly 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  margin-right: 8px;
}

.c1:hover {
  cursor: pointer;
}

.c2 {
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid #4718b4;
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: 8px;
  -webkit-transition: background-color 0.3s ease-out;
  transition: background-color 0.3s ease-out;
}

.c0:hover .c2 {
  background-color: #ece7f7;
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  />
</div>
`;

exports[`<Checkbox /> renders correctly after check 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  margin-right: 8px;
}

.c1:hover {
  cursor: pointer;
}

.c0:hover .c3 {
  background-color: #ece7f7;
}

.c2 {
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid #4718b4;
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: 8px;
  -webkit-transition: background-color 0.3s ease-out;
  transition: background-color 0.3s ease-out;
  background-color: #4718b4;
  background-image: url(check.svg);
  background-repeat: no-repeat;
  background-position: center;
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  />
</div>
`;

exports[`<Checkbox /> renders correctly with defaultChecked 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  margin-right: 8px;
}

.c1:hover {
  cursor: pointer;
}

.c0:hover .c3 {
  background-color: #ece7f7;
}

.c2 {
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid #4718b4;
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: 8px;
  -webkit-transition: background-color 0.3s ease-out;
  transition: background-color 0.3s ease-out;
  background-color: #4718b4;
  background-image: url(check.svg);
  background-repeat: no-repeat;
  background-position: center;
}

<div
  class="c0 c1"
>
  <div
    checked=""
    class="c2"
  />
</div>
`;

exports[`<Checkbox /> renders correctly with disabled 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  margin-right: 8px;
  opacity: 0.5;
}

.c1:hover {
  cursor: pointer;
}

.c1:hover {
  cursor: default;
}

.c0:hover .c3 {
  background-color: #ece7f7;
}

.c2 {
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid #4718b4;
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: 8px;
  -webkit-transition: background-color 0.3s ease-out;
  transition: background-color 0.3s ease-out;
}

<div
  class="c0 c1"
  disabled=""
>
  <div
    class="c2"
    disabled=""
  />
</div>
`;

exports[`<Checkbox /> renders correctly with label 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  margin-right: 8px;
}

.c1:hover {
  cursor: pointer;
}

.c3 {
  display: inline-block;
  font-size: 16px;
}

.c2 {
  display: inline-block;
  width: 18px;
  height: 18px;
  min-width: 18px;
  border: 2px solid #4718b4;
  border-radius: 2px;
  box-sizing: border-box;
  margin-right: 8px;
  -webkit-transition: background-color 0.3s ease-out;
  transition: background-color 0.3s ease-out;
}

.c0:hover .c2 {
  background-color: #ece7f7;
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  />
  <div
    class="c3"
  >
    Label
  </div>
</div>
`;
